Ipv4 ipv6

sanderjo opened the pull request.

sanderjo commented:
Improvements with newsserver load balancing aka happy eyeballs. 
Most important one: parallel happyeyeballs (multithreaded), and only wait for the quickest successful connect. Example below:

```
2015-09-25 21:33:58,603::INFO::[downloader:423] 1@newszilla.xs4all.nl: Initiating connection
2015-09-25 21:33:58,662::INFO::[happyeyeballs:115] Quickest IP address for newszilla.xs4all.nl (port 119, ssl 0, preferipv6 True) is 2001:888:0:18::119
2015-09-25 21:33:58,663::DEBUG::[happyeyeballs:117] Happy Eyeballs lookup took 2.856 microseconds
2015-09-25 21:33:58,663::DEBUG::[downloader:112] For server newszilla.xs4all.nl, using IP 2001:888:0:18::119 as server


2015-09-25 21:34:02,714::INFO::[downloader:423] 1@news.thundernews.com: Initiating connection
2015-09-25 21:34:02,786::INFO::[happyeyeballs:115] Quickest IP address for news.thundernews.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::146
2015-09-25 21:34:02,787::DEBUG::[happyeyeballs:117] Happy Eyeballs lookup took 3.986 microseconds
2015-09-25 21:34:02,787::DEBUG::[downloader:112] For server news.thundernews.com, using IP 2001:4de0:401:7::146 as server
```

sanderjo commented:
Interesting lines from sabnzbd.log with a lot of IPv6 capable newsservers filled out in SAB:
```
2015-09-26 10:01:14,936::INFO::[happyeyeballs:115] Quickest IP address for eu.z51.biz (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::222
2015-09-26 10:01:15,038::INFO::[happyeyeballs:115] Quickest IP address for eu.z51.biz (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::236
2015-09-26 10:01:15,141::INFO::[happyeyeballs:115] Quickest IP address for eu.z51.biz (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::222
2015-09-26 10:01:15,273::INFO::[happyeyeballs:115] Quickest IP address for eu.z51.biz (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::222
2015-09-26 10:01:15,394::INFO::[happyeyeballs:115] Quickest IP address for news.prepaid-usenet.de (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::222
2015-09-26 10:01:15,503::INFO::[happyeyeballs:115] Quickest IP address for news.prepaid-usenet.de (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::236
2015-09-26 10:01:15,609::INFO::[happyeyeballs:115] Quickest IP address for news.prepaid-usenet.de (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::236
2015-09-26 10:01:15,735::INFO::[happyeyeballs:115] Quickest IP address for news.prepaid-usenet.de (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::236
2015-09-26 10:01:16,026::INFO::[happyeyeballs:115] Quickest IP address for secure.newsgroup-binaries.com (port 119, ssl 0, preferipv6 True) is 69.16.179.22
2015-09-26 10:01:16,199::INFO::[happyeyeballs:115] Quickest IP address for secure.newsgroup-binaries.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:2101:119e::23
2015-09-26 10:01:16,437::INFO::[happyeyeballs:115] Quickest IP address for secure.newsgroup-binaries.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:2101:119e::23
2015-09-26 10:01:16,620::INFO::[happyeyeballs:115] Quickest IP address for secure.newsgroup-binaries.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:2101:119e::23
2015-09-26 10:01:16,926::INFO::[happyeyeballs:115] Quickest IP address for secure.thenewsgroups.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:2101:119e::22
2015-09-26 10:01:17,126::INFO::[happyeyeballs:115] Quickest IP address for secure.thenewsgroups.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:2101:119e::22
2015-09-26 10:01:17,384::INFO::[happyeyeballs:115] Quickest IP address for secure.thenewsgroups.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:2101:119e::22
2015-09-26 10:01:17,597::INFO::[happyeyeballs:115] Quickest IP address for secure.thenewsgroups.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:2101:119e::23
2015-09-26 10:01:17,783::INFO::[happyeyeballs:115] Quickest IP address for news.newsgroupdirect.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::146
2015-09-26 10:01:17,951::INFO::[happyeyeballs:115] Quickest IP address for news.newsgroupdirect.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::146
2015-09-26 10:01:18,222::INFO::[happyeyeballs:115] Quickest IP address for news.newsgroupdirect.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::147
2015-09-26 10:01:18,373::INFO::[happyeyeballs:115] Quickest IP address for news.newsgroupdirect.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::146
2015-09-26 10:01:18,487::INFO::[happyeyeballs:115] Quickest IP address for secure.eu.thundernews.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::146
2015-09-26 10:01:18,714::INFO::[happyeyeballs:115] Quickest IP address for secure.eu.thundernews.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::147
2015-09-26 10:01:18,837::INFO::[happyeyeballs:115] Quickest IP address for secure.eu.thundernews.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::147
2015-09-26 10:01:18,977::INFO::[happyeyeballs:115] Quickest IP address for secure.eu.thundernews.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::147
2015-09-26 10:01:19,147::INFO::[happyeyeballs:115] Quickest IP address for news.thundernews.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::146
2015-09-26 10:01:19,310::INFO::[happyeyeballs:115] Quickest IP address for news.thundernews.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::147
2015-09-26 10:01:19,442::INFO::[happyeyeballs:115] Quickest IP address for news.thundernews.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::146
2015-09-26 10:01:19,589::INFO::[happyeyeballs:115] Quickest IP address for news.thundernews.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::147
2015-09-26 10:01:19,698::INFO::[happyeyeballs:115] Quickest IP address for news.iload-usenet.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::222
2015-09-26 10:01:19,812::INFO::[happyeyeballs:115] Quickest IP address for news.iload-usenet.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::222
2015-09-26 10:01:19,931::INFO::[happyeyeballs:115] Quickest IP address for news.iload-usenet.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::222
2015-09-26 10:01:20,061::INFO::[happyeyeballs:115] Quickest IP address for news.iload-usenet.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::222
2015-09-26 10:01:20,189::INFO::[happyeyeballs:115] Quickest IP address for eu.newsgroup-binaries.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::222
2015-09-26 10:01:20,305::INFO::[happyeyeballs:115] Quickest IP address for eu.newsgroup-binaries.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::222
2015-09-26 10:01:20,426::INFO::[happyeyeballs:115] Quickest IP address for eu.newsgroup-binaries.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::236
2015-09-26 10:01:20,559::INFO::[happyeyeballs:115] Quickest IP address for eu.newsgroup-binaries.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::236
2015-09-26 10:01:20,695::INFO::[happyeyeballs:115] Quickest IP address for news.xs4all.nl (port 119, ssl 0, preferipv6 True) is 2001:888:0:4::119
2015-09-26 10:01:20,815::INFO::[happyeyeballs:115] Quickest IP address for news.xs4all.nl (port 119, ssl 0, preferipv6 True) is 2001:888:0:4::119
2015-09-26 10:01:20,931::INFO::[happyeyeballs:115] Quickest IP address for news.xs4all.nl (port 119, ssl 0, preferipv6 True) is 2001:888:0:4::119
2015-09-26 10:01:21,023::INFO::[happyeyeballs:115] Quickest IP address for news.xs4all.nl (port 119, ssl 0, preferipv6 True) is 2001:888:0:4::119
2015-09-26 10:01:21,135::INFO::[happyeyeballs:115] Quickest IP address for news.usenetnewsserver.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::147
2015-09-26 10:01:21,251::INFO::[happyeyeballs:115] Quickest IP address for news.usenetnewsserver.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::146
2015-09-26 10:01:21,381::INFO::[happyeyeballs:115] Quickest IP address for news.usenetnewsserver.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::146
2015-09-26 10:01:21,479::INFO::[happyeyeballs:115] Quickest IP address for news.usenetnewsserver.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::147
2015-09-26 10:01:21,781::INFO::[happyeyeballs:115] Quickest IP address for us.newsgroupdirect.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:2101:119e::22
2015-09-26 10:01:21,947::INFO::[happyeyeballs:115] Quickest IP address for us.newsgroupdirect.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:2101:119e::23
2015-09-26 10:01:22,115::INFO::[happyeyeballs:115] Quickest IP address for us.newsgroupdirect.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:2101:119e::22
2015-09-26 10:01:22,420::INFO::[happyeyeballs:115] Quickest IP address for us.newsgroupdirect.com (port 119, ssl 0, preferipv6 True) is 69.16.179.22
2015-09-26 10:01:22,498::INFO::[happyeyeballs:115] Quickest IP address for news.xlusenet.nl (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::222
2015-09-26 10:01:22,590::INFO::[happyeyeballs:115] Quickest IP address for news.xlusenet.nl (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::236
2015-09-26 10:01:22,667::INFO::[happyeyeballs:115] Quickest IP address for news.xlusenet.nl (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::222
2015-09-26 10:01:22,746::INFO::[happyeyeballs:115] Quickest IP address for news.xlusenet.nl (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::236
2015-09-26 10:01:22,841::INFO::[happyeyeballs:115] Quickest IP address for reader.snelnl.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::147
2015-09-26 10:01:22,932::INFO::[happyeyeballs:115] Quickest IP address for reader.snelnl.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::146
2015-09-26 10:01:23,028::INFO::[happyeyeballs:115] Quickest IP address for reader.snelnl.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::147
2015-09-26 10:01:23,150::INFO::[happyeyeballs:115] Quickest IP address for reader.snelnl.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::146
2015-09-26 10:01:23,240::INFO::[happyeyeballs:115] Quickest IP address for news.shemes.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::236
2015-09-26 10:01:23,316::INFO::[happyeyeballs:115] Quickest IP address for news.shemes.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::236
2015-09-26 10:01:23,386::INFO::[happyeyeballs:115] Quickest IP address for news.shemes.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::236
2015-09-26 10:01:23,460::INFO::[happyeyeballs:115] Quickest IP address for news.shemes.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::236
2015-09-26 10:01:23,562::INFO::[happyeyeballs:115] Quickest IP address for news.alibis.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::146
2015-09-26 10:01:24,709::INFO::[happyeyeballs:115] Quickest IP address for news.alibis.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::146
2015-09-26 10:01:24,832::INFO::[happyeyeballs:115] Quickest IP address for news.alibis.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::146
2015-09-26 10:01:24,961::INFO::[happyeyeballs:115] Quickest IP address for news.alibis.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::147
2015-09-26 10:01:25,138::INFO::[happyeyeballs:115] Quickest IP address for secure.us.thundernews.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:2101:119e::23
2015-09-26 10:01:25,417::INFO::[happyeyeballs:115] Quickest IP address for secure.us.thundernews.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:2101:119e::22
2015-09-26 10:01:25,626::INFO::[happyeyeballs:115] Quickest IP address for secure.us.thundernews.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:2101:119e::23
2015-09-26 10:01:25,925::INFO::[happyeyeballs:115] Quickest IP address for secure.us.thundernews.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:2101:119e::23
```


For easier reading: everything up to "Quickest IP address for " removed:
```
eu.z51.biz (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::222
eu.z51.biz (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::236
eu.z51.biz (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::222
eu.z51.biz (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::222
news.prepaid-usenet.de (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::222
news.prepaid-usenet.de (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::236
news.prepaid-usenet.de (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::236
news.prepaid-usenet.de (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::236
secure.newsgroup-binaries.com (port 119, ssl 0, preferipv6 True) is 69.16.179.22
secure.newsgroup-binaries.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:2101:119e::23
secure.newsgroup-binaries.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:2101:119e::23
secure.newsgroup-binaries.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:2101:119e::23
secure.thenewsgroups.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:2101:119e::22
secure.thenewsgroups.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:2101:119e::22
secure.thenewsgroups.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:2101:119e::22
secure.thenewsgroups.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:2101:119e::23
news.newsgroupdirect.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::146
news.newsgroupdirect.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::146
news.newsgroupdirect.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::147
news.newsgroupdirect.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::146
secure.eu.thundernews.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::146
secure.eu.thundernews.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::147
secure.eu.thundernews.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::147
secure.eu.thundernews.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::147
news.thundernews.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::146
news.thundernews.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::147
news.thundernews.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::146
news.thundernews.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::147
news.iload-usenet.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::222
news.iload-usenet.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::222
news.iload-usenet.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::222
news.iload-usenet.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::222
eu.newsgroup-binaries.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::222
eu.newsgroup-binaries.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::222
eu.newsgroup-binaries.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::236
eu.newsgroup-binaries.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::236
news.xs4all.nl (port 119, ssl 0, preferipv6 True) is 2001:888:0:4::119
news.xs4all.nl (port 119, ssl 0, preferipv6 True) is 2001:888:0:4::119
news.xs4all.nl (port 119, ssl 0, preferipv6 True) is 2001:888:0:4::119
news.xs4all.nl (port 119, ssl 0, preferipv6 True) is 2001:888:0:4::119
news.usenetnewsserver.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::147
news.usenetnewsserver.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::146
news.usenetnewsserver.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::146
news.usenetnewsserver.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::147
us.newsgroupdirect.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:2101:119e::22
us.newsgroupdirect.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:2101:119e::23
us.newsgroupdirect.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:2101:119e::22
us.newsgroupdirect.com (port 119, ssl 0, preferipv6 True) is 69.16.179.22
news.xlusenet.nl (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::222
news.xlusenet.nl (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::236
news.xlusenet.nl (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::222
news.xlusenet.nl (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::236
reader.snelnl.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::147
reader.snelnl.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::146
reader.snelnl.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::147
reader.snelnl.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::146
news.shemes.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::236
news.shemes.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::236
news.shemes.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::236
news.shemes.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:1::236
news.alibis.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::146
news.alibis.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::146
news.alibis.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::146
news.alibis.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:401:7::147
secure.us.thundernews.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:2101:119e::23
secure.us.thundernews.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:2101:119e::22
secure.us.thundernews.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:2101:119e::23
secure.us.thundernews.com (port 119, ssl 0, preferipv6 True) is 2001:4de0:2101:119e::23
```
So only two times IPv4 is preferred:
```
secure.newsgroup-binaries.com (port 119, ssl 0, preferipv6 True) is 69.16.179.22
us.newsgroupdirect.com (port 119, ssl 0, preferipv6 True) is 69.16.179.22
```

shypike commented on sabnzbd/utils/happyeyeballs.py:
```diff
@@ -16,62 +73,76 @@ def happyeyeballs(HOST, **kwargs):
+        preferipv6=kwargs['preferipv6']
     except:
-        DEBUG=False
+        preferipv6=True	# prefer IPv6, so give IPv6 connects a head start by delaying IPv4
```
Please stick to PEP8 and use:
DEBUG = False
preferipv6 = True

sanderjo replied:
Done:
https://github.com/sanderjo/sabnzbd/commit/9cfc767d8e22ca2791f5c757b65817f92b3bc634


On Sat, Sep 26, 2015 at 3:17 PM, shypike <notifications@github.com> wrote:

> In sabnzbd/utils/happyeyeballs.py
> <https://github.com/sabnzbd/sabnzbd/pull/340#discussion_r40493498>:
>
> >      except:
> > -        DEBUG=False
> > +        preferipv6=True	# prefer IPv6, so give IPv6 connects a head start by delaying IPv4
>
> Please stick to PEP8 and use:
> DEBUG = False
> preferipv6 = True
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sabnzbd/sabnzbd/pull/340/files#r40493498>.
>



--

shypike commented:
The code crashes, because server names are (or can be) in Unicode. For Unicode the translate() function takes just one parameter, but for ASCII it takes two.
Also you've added tabs to the newswrapper.py file, please use only spaces.

sanderjo commented:
FWIW: I can't reproduce the problem in SAB, but I can in on the python command line. So: replaced with a regexp.
Plus a dozen tabs replaced by spaces

Safihre commented:
As thezoggy would say: "squash and rebase", so the commit history stays clean. 
Took me a while to figure out how, so in case, I'm just sharing the knowledge. Put in git shell:
git rebase -i HEAD~3
(now a text editor window will come up, there you replace the 'pick' by 'f' in front of the last 2 commits, save and close the window). 
When it's done:
git push -f
This will make the last 3 commits into 1!

thezoggy commented:
TortoiseGIT makes it very easy to do all this on windows.. nice gui/right click do what you want..

sanderjo commented:
After ```$ git rebase -i HEAD~5``` I changed four pick lines to "f":

```
pick 8ca00a9 Parallel HappyEyeballs and Solve issue 339
f a667f09 Other message in case of no AAAA resolving by DNS
f c8a7e5b if Load-Balancing-with-IPv6 is set (2), we can ignore the EXTERNAL_IPv6
f 9cfc767 PEP 0008if Load-Balancing-with-IPv6 is set (2), we can ignore the EXTERNAL_IPv6
f 334a11c no more translate (because of unicode problems), but a regexp. Plus more PEP0008

# Rebase bcf5343..334a11c onto bcf5343 (5 command(s))
```
Exit and save:
```
$ git rebase -i HEAD~5
[detached HEAD 7aed4d8] Parallel HappyEyeballs and Solve issue 339
 Date: Fri Sep 25 21:13:17 2015 +0200
 3 files changed, 173 insertions(+), 82 deletions(-)
 rewrite sabnzbd/utils/happyeyeballs.py (73%)
Successfully rebased and updated refs/heads/ipv4-ipv6.
```
And then:
```
$ git push -f
Username for 'https://github.com': sanderjo
Password for 'https://sanderjo@github.com': 
Counting objects: 7, done.
Delta compression using up to 2 threads.
Compressing objects: 100% (7/7), done.
Writing objects: 100% (7/7), 2.95 KiB | 0 bytes/s, done.
Total 7 (delta 5), reused 0 (delta 0)
To https://github.com/sanderjo/sabnzbd.git
 + 334a11c...7aed4d8 ipv4-ipv6 -> ipv4-ipv6 (forced update)
```
@shypike Did it automatically arrive in the existing pull request, or do I have to send a new pull request?

shypike commented:
@sanderjo I still see a pull-request that has conflicts. So probably it wasn't enough.

sanderjo commented:
No, I didn't sync upstream. Is that still possible (if so: how), or is it lost now?

shypike commented:
@sanderjo Good question, I'm always confused by that too.
The easiest fix is to create a new clone and cherrypick your commit to the new clone.

thezoggy commented:
@sanderjo sure you can rebase and force push. since no one has code that is built off your changes..

Safihre commented:
git fetch upstream 
git rebase upstream/develop 

But I wonder if it will give conflicts as well, since Git now has a problem. Anyway, the commit that fixes the conflicts you can then just rebase together again. 
If you then force push it will be updated in this pull request as well!

shypike commented:
Already merged through other pull request.
